src/invariant/OrderBookIsNotCrossed.cpp

```diff
@@ -0,0 +1,162 @@
+// Copyright 2019 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "invariant/OrderBookIsNotCrossed.h"
+#include "invariant/InvariantManager.h"
+#include "ledger/LedgerTxn.h"
+#include "lib/util/format.h"
+#include "main/Application.h"
+#include "xdr/Stellar-ledger-entries.h"
+
+#include <numeric>
+
+namespace stellar
+{
+AssetId
+getAssetID(Asset const& asset)
+{
+    auto r = std::string{};
+    switch (asset.type())
+    {
+    case stellar::ASSET_TYPE_NATIVE:
+        return std::string{"XLM-native"};
+    case stellar::ASSET_TYPE_CREDIT_ALPHANUM4:
+        assetCodeToStr(asset.alphaNum4().assetCode, r);
+        return r + "-" + KeyUtils::toStrKey(getIssuer(asset));
+    case stellar::ASSET_TYPE_CREDIT_ALPHANUM12:
+        assetCodeToStr(asset.alphaNum12().assetCode, r);
+        return r + "-" + KeyUtils::toStrKey(getIssuer(asset));
```

Comment:
I think you need to add a `default` to avoid compiler warnings, even though all cases are covered here. Usually it's an `abort` call, see `BasicWork::getStatus()` as an example